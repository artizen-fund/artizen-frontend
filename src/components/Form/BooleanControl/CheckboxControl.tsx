import React from 'react'
import styled from 'styled-components'
import { breakpoint, palette } from '@theme'
import { rgba } from '@lib'
import { BooleanControlProps } from './'

const Checkbox = ({
  outline = false,
  required,
  label,
  data,
  handleChange,
  path,
  disabled = false,
}: BooleanControlProps) => {
  const small = false
  return (
    <Wrapper {...{ disabled, small }}>
      <Box {...{ small }}>
        <Input type="checkbox" required={!!required} onChange={_ => handleChange(path, !data)} />
        <Checkmark {...{ outline, small }} />
      </Box>
      <Label>{typeof label === 'object' ? label[0] : label}</Label>
    </Wrapper>
  )
}

const Wrapper = styled.label<{ disabled: boolean; small?: boolean }>`
  position: relative;
  display: flex;
  align-items: ${props => (props.small ? 'flex-start' : 'center')};
  padding: ${props => (props.small ? 0 : 16)}px 0;
  user-select: none;
  cursor: ${props => (props.disabled ? 'not-allowed' : 'pointer')};
  opacity: ${props => (props.disabled ? 0.32 : 1)};

  &:hover input ~ span {
    ${rgba(palette.moon, 0.8)};
    @media (prefers-color-scheme: dark) {
      ${rgba(palette.night, 0.8)};
    }
  }

  & input:checked ~ span {
    background-color: ${rgba(palette.moon)};
    @media (prefers-color-scheme: dark) {
      background-color: ${rgba(palette.night)};
    }

    &:after {
      transform: rotate(45deg) scale3d(1, 1, 1);
      opacity: 1;
    }
  }
`

const Box = styled.div<{ small?: boolean }>`
  flex: 0 0 auto;
  position: relative;
  width: ${props => (props.small ? '16px' : '24px')};
  height: ${props => (props.small ? '16px' : '24px')};
  margin-right: ${props => (props.small ? '8px' : '16px')};

  @media only screen and (min-width: ${breakpoint.laptop}px) {
    width: 32px;
    height: 32px;
    margin-right: 24px;
  }

  @media only screen and (min-width: ${breakpoint.desktop}px) {
    width: 40px;
    height: 40px;
  }
`

const Input = styled.input`
  position: absolute;
  opacity: 0;
  height: 0;
  width: 0;
`

const Checkmark = styled.span<{ outline: boolean; small?: boolean }>`
  position: absolute;
  top: 0;
  left: 0;
  background-color: ${rgba(palette.moon, 0)};
  width: 100%;
  height: 100%;
  border: 1px solid ${props => (props.outline ? rgba(palette.night) : rgba(palette.moon))};
  @media (prefers-color-scheme: dark) {
    border: 1px solid ${props => (props.outline ? rgba(palette.moon) : rgba(palette.night))};
  }
  border-radius: 50%;
  ${props =>
    !props.outline &&
    `
    box-shadow: 0.5px 1px 0px rgba(0, 0, 0, 0.16);
  `}
  appearance: none;
  transition: background-color 0.25s ease-in-out;

  &:after {
    position: absolute;
    border-style: solid;
    border-color: ${rgba(palette.night)};
    @media (prefers-color-scheme: dark) {
      border-color: ${rgba(palette.moon)};
    }
    ${props =>
      props.small
        ? `
      left: calc(50% - 1px);
      top: calc(50% - 4px);
      width: 3px;
      height: 7px;
      border-width: 0 1.5px 1px 0;
    `
        : `
      left: calc(50% - 1.5px) !important;
      top: calc(50% - 3.75px) !important;
      width: 3px !important;
      height: 6px !important;
      border-width: 0 1.125px 1.125px 0 !important;
    `}
    display: block;
    transform: rotate(45deg) scale3d(0, 0, 1);
    opacity: 0;
    content: '';
    transition: opacity 0.25s ease-in-out, transform 0.6s cubic-bezier(0.44, 1.86, 0.74, 1);

    @media only screen and (min-width: ${breakpoint.laptop}px) {
      left: calc(50% - 3px);
      top: calc(50% - 7.5px);
      width: 6px;
      height: 12px;
      border-width: 0 2.5px 2.5px 0;
    }
  }
`

const Label = styled.span`
  display: block;
  max-width: 280px;

  & a {
    position: relative;
    display: inline-flex;
    margin: 0 1px;
    color: ${rgba(palette.coral)};
    overflow: hidden;

    &::after {
      z-index: 2;
      position: absolute;
      bottom: 0;
      left: 0;
      background-color: ${rgba(palette.coral, 0.64)};
      width: 100%;
      height: 1px;
      transform: translate3d(-100%, 0, 0);
      transition: color 0.25s ease-in-out, transform 0.35s ease-in-out;
      content: '';
      pointer-events: none;
    }

    &:hover {
      &::after {
        transform: translate3d(0, 0, 0);
      }
    }
  }

  @media only screen and (min-width: ${breakpoint.laptop}px) {
    max-width: 300px;
  }

  @media only screen and (min-width: ${breakpoint.desktop}px) {
    max-width: 320px;
  }
`

export default Checkbox
