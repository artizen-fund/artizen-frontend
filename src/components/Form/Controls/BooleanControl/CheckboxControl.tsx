import styled from 'styled-components'
import { breakpoint, palette, typography } from '@theme'
import { rgba } from '@lib'
import { BooleanControlProps } from './'

const Checkbox = ({
  required,
  label,
  inverted,
  data,
  handleChange,
  path,
  enabled = true,
  uischema,
  ...props
}: BooleanControlProps) => (
  <Wrapper gridArea={path} {...{ inverted, enabled }} {...props} id={uischema?.scope}>
    <Box>
      <Input
        type="checkbox"
        required={!!required}
        onChange={_ => handleChange(path, !data)}
        checked={data}
        disabled={!enabled}
      />
      <Checkmark {...{ inverted }} />
    </Box>
    <Label {...{ inverted, enabled }}>{typeof label === 'object' ? label[0] : label}</Label>
  </Wrapper>
)

const Box = styled.div`
  flex: 0 0 auto;
  position: relative;
  width: 24px;
  height: 24px;

  @media only screen and (min-width: ${breakpoint.laptop}px) {
    width: 32px;
    height: 32px;
  }

  @media only screen and (min-width: ${breakpoint.desktop}px) {
    width: 40px;
    height: 40px;
  }
`

const Input = styled.input`
  position: absolute;
  opacity: 0;
  height: 0;
  width: 0;
`

const Checkmark = styled.span<Pick<BooleanControlProps, 'inverted'>>`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;

  border-radius: 9999px;
  appearance: none;
  transition: background-color 0.25s ease-in-out, box-shadow 0.15s ease-in-out;

  background-color: ${props => rgba(props.inverted ? palette.night : palette.white)};
  border: 2px solid ${props => rgba(props.inverted ? palette.white : palette.slate)};
  @media (prefers-color-scheme: dark) {
    background-color: ${rgba(palette.slate)};
    border-color ${rgba(palette.moon)};
  }
  /* see [Wrapper -> & input ~ span] for state style changes */
  
  &:after {
    position: absolute;
    border-style: solid;
    border-color: ${props => rgba(props.inverted ? palette.night : palette.white)};
    @media (prefers-color-scheme: dark) {
      border-color ${rgba(palette.night)};
    }
    left: calc(50% - 1px);
    top: calc(50% - 4px);
    width: 3px;
    height: 7px;
    border-width: 0 1.5px 1px 0;
    display: block;
    transform: rotate(45deg) scale3d(0, 0, 1);
    opacity: 0;
    content: '';
    transition: opacity 0.25s ease-in-out, transform 0.6s cubic-bezier(0.44, 1.86, 0.74, 1);
  }
`

const Label = styled.span<Pick<BooleanControlProps, 'enabled' | 'inverted'>>`
  display: block;
  ${typography.label.l2}
  color: ${props => rgba(!props.enabled ? palette.barracuda : props.inverted ? palette.white : palette.night)};
  @media (prefers-color-scheme: dark) {
    color: ${props => rgba(!props.enabled ? palette.barracuda : palette.white)};
  }
`

const Wrapper = styled.label<Pick<BooleanControlProps, 'enabled' | 'inverted'> & { gridArea?: string }>`
  position: relative;
  ${props => props.gridArea && `grid-area: ${props.gridArea};`}
  display: flex;
  flex-direction: row;
  align-items: center;
  padding: 16px 0;
  gap: 8px;
  @media only screen and (min-width: ${breakpoint.laptop}px) {
    gap: 10px;
  }
  @media only screen and (min-width: ${breakpoint.desktop}px) {
    gap: 12px;
  }

  user-select: none;

  cursor: ${props => (!props.enabled ? 'not-allowed' : 'pointer')};
  pointer-events: ${props => (!props.enabled ? 'none' : 'all')};

  & input:checked ~ span {
    background-color: ${props => rgba(props.inverted ? palette.white : palette.slate)};
    @media (prefers-color-scheme: dark) {
      background-color: ${rgba(palette.moon)};
    }

    &:after {
      transform: rotate(45deg) scale3d(1, 1, 1);
      opacity: 1;
    }
  }

  & input:checked:disabled ~ span,
  & input:disabled ~ span {
    background-color: ${props => (props.inverted ? rgba(palette.barracuda, 0.4) : rgba(palette.stone))};
    border: 1px solid ${props => (props.inverted ? rgba(palette.barracuda, 0.4) : rgba(palette.stone))};
    &::after {
      border-color: ${rgba(palette.barracuda)};
      transform: rotate(45deg) scale3d(1, 1, 1);
      opacity: 1;
    }
    @media (prefers-color-scheme: dark) {
      background-color: ${rgba(palette.barracuda, 0.4)};
      border: 1px solid ${rgba(palette.barracuda, 0.4)};
      &::after {
        border-color: ${rgba(palette.barracuda)};
        transform: rotate(45deg) scale3d(1, 1, 1);
        opacity: 1;
      }
    }
  }

  &:hover ${Checkmark} {
    box-shadow: 0px 4px 16px rgba(0, 0, 0, 0.16);
  }
`

export default Checkbox
