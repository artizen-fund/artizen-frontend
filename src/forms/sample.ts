export const schema = {
  type: 'object',
  properties: {
    name: {
      type: 'string',
      minLength: 3,
      description: 'Please enter your name',
    },
    vegetarian: {
      type: 'boolean',
    },
    vegetarianSure: {
      type: 'boolean',
    },
    carnivore: {
      type: 'boolean',
    },
    birthDate: {
      type: 'string',
      format: 'date',
    },
    nationality: {
      type: 'string',
      enum: ['', 'DE', 'IT', 'JP', 'US', 'RU', 'Other'],
    },
    iceCream: {
      type: 'string',
      enum: ['vanilla', 'chocolate', 'strawberry', 'neapolitan'],
    },
    personalData: {
      type: 'object',
      properties: {
        age: {
          type: 'integer',
          description: 'Please enter your age.',
        },
        height: {
          type: 'number',
        },
        drivingSkill: {
          type: 'number',
          maximum: 10,
          minimum: 1,
          default: 7,
        },
      },
      required: ['age', 'height'],
    },
    occupation: {
      type: 'string',
    },
    postalCode: {
      type: 'string',
      maxLength: 5,
    },
  },
  required: ['occupation', 'nationality'],
}

export const uischema = {
  type: 'VerticalLayout',
  elements: [
    {
      type: 'HorizontalLayout',
      elements: [
        {
          type: 'Control',
          scope: '#/properties/name',
        },
        {
          type: 'Control',
          scope: '#/properties/vegetarian',
        },
        {
          type: 'Control',
          scope: '#/properties/vegetarianSure',
          options: {
            format: 'checkbox',
            size: 'small',
          },
        },
        {
          type: 'Control',
          scope: '#/properties/carnivore',
          options: {
            format: 'switch',
          },
        },
        {
          type: 'Control',
          scope: '#/properties/iceCream',
          options: {
            format: 'segmented',
          },
        },
        {
          type: 'Control',
          scope: '#/properties/personalData/properties/age',
        },
        {
          type: 'Control',
          scope: '#/properties/birthDate',
        },
      ],
    },
    {
      type: 'Label',
      text: 'Additional Information',
    },
    {
      type: 'HorizontalLayout',
      elements: [
        {
          type: 'Control',
          scope: '#/properties/personalData/properties/height',
        },
        {
          type: 'Control',
          scope: '#/properties/nationality',
        },
        {
          type: 'Control',
          scope: '#/properties/occupation',
          suggestion: [
            'Accountant',
            'Engineer',
            'Freelancer',
            'Journalism',
            'Physician',
            'Student',
            'Teacher',
            'Other',
          ],
        },
      ],
    },
  ],
}

export const initialState = {
  name: undefined,
  vegetarian: undefined,
  vegetarianSure: undefined,
  carnivore: undefined,
  birthDate: undefined,
  iceCream: undefined,
  personalData: {
    age: undefined,
  },
  postalCode: undefined,
}
