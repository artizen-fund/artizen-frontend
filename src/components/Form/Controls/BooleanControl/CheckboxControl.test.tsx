import '@testing-library/jest-dom/extend-expect'
import { render } from '@testing-library/react'
import CheckboxControl from './CheckboxControl'
import { schema, uischema } from '@forms/sample'

describe('CheckboxControl', () => {
  const INPUT_PATH = '#/properties/booleanCheckExample'

  it('renders SelectControl unchanged', () => {
    const handleChange = () => console.log('derp')
    const { container } = render(
      <CheckboxControl
        data="vanilla"
        path={uischema.elements.find(e => e.scope === INPUT_PATH)?.scope || 'some-made-up-scope'}
        {...{ handleChange, schema, uischema }}
        label="Checkbox control test"
      />,
    )
    expect(container).toMatchSnapshot()
  })
})

/* 
Possible future testsâ€¦

- onchange triggers console log
*/
