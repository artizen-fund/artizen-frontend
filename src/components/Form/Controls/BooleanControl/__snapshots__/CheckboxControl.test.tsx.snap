// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxControl renders SelectControl unchanged 1`] = `
<div>
  <label
    class="sc-dd35d48e-4 fBdSSs"
  >
    <div
      class="sc-dd35d48e-0 lmHjtw"
    >
      <input
        checked=""
        class="sc-dd35d48e-1 clTMzK"
        type="checkbox"
      />
      <span
        class="sc-dd35d48e-2 dtceeR"
      />
    </div>
    <span
      class="sc-dd35d48e-3 eAaGSv"
    >
      Checkbox control test
    </span>
  </label>
</div>
`;
